examples/smqtk/show_smqtk.py

```diff
@@ -0,0 +1,23 @@
+"""Meta-configuration demonstration."""
+
+from tinker import protocol
+from smqtk_core import Configurable
+
+
+class ShowSmqtk(protocol.Protocol):
+    """
+    A protocol designed for testing SMQTK integration.
+
+    Will initialize and then print out config / instance details
+    """
+
+    def get_config(self):
+        """Return a default configuration dictionary."""
+        return {}
+
+    def run_protocol(self, config):
```

Comment:
The main value add of the Protocol class is to provide a uniform runtime interface for Tinker Engine itself. We can certainly do Pythonic magic with individual Protocol implementations if that's helpful to individual protocol writers.

The `config` argument passed into `run_protocol()` comes from the configuration handler. It's not intended to be a SMQTK configuration object, but rather a generic configuration for the experimental setup. I am not sure that it makes sense to leverage SMQTK's config mechanism for this--in particular, the `smqtk` directive manufactures SMQTK objects that can be used by the `run_protocol()` method, so this is more of a general Python environment for running the experiment, which is kicked off by a single configuration supplied by the Tinker Engine runtime.

I hope that makes sense--let me know if not.

Reply:
> The main value add of the Protocol class is to provide a uniform runtime interface for Tinker Engine itself.

:+1: 100%

> ... but rather a generic configuration for the experimental setup.

My concern with passing generic configuration for the experimental setup is that a Protocol implementation may not know to care about this or create a requirement on the contents of the config that either aren't real, protocol-implementation-specific, or are not reactive to changes in structure of the passed config going forward. Apologies if this is documented somewhere an I have missed reading it: Is there standard content being passed in this `run_protocol` config? If so, then why not make a concrete type or dataclass for this content? To my understanding, that would indicate there is specific structure imposed in this argument, also leading to mypy typechecking errors if the now-structured config is utilized outside of it's intended scope.

> I am not sure that it makes sense to leverage SMQTK's config mechanism for this

My perspective here is that the protocol, for one derived configuration (e.g. after doing any iteration expansion, etc.) is constructed once and then has its `run_protocol(...)` invoked once.
I also imagine that experimental setup is defined and dictated by the specific protocol implementation being used for the experimental runs.
If that is the case, then experimental setup, as I imagine is dictated by the specific Protocol implementation being used for the run, would occur once per Protocol implementation instance and within the protocol instance itself?
Is the concern or desire to use the `run_protocol` config one of utilizing shared resources across/between Protocol implementation runs of different configurations?

A glimpse as to what's in my head for SMQTK-like configuration, modulo reality and other tinker configuration features, to have as a strawman to point at:
```yml
tinker_setting_1: foo
tinker_setting_2: bar
protocol:  # contents here maybe derivable via `smqtk_core.configuration.make_default_config`
  type: ExperimentProtocolA  # pick you Protocol implementation instance here.
  ExperimentProtocolA:  # Some protocol implementation defined somewhere discoverable
    setup_param_1:
      iterate:  # a new thing I saw in this PR could be used here
        - 0
        - 1
    loops: 10
    descriptor_generator:  # A nested "Plugfigurable"
      # Maybe even iterate on nested implementations?
      type:
        iterate:
          - smqtk_torch.TorchDescriptorGenerator  # I'm making up this plugin and package
          - smqtk_descriptors.impls.descriptor_generator.caffe1.CaffeDescriptorGenerator
        smqtk_torch.TorchDescriptorGenerator:
          implementation_parameter: ...
        smqtk_descriptors.impls.descriptor_generator.caffe1.CaffeDescriptorGenerator:
          implementation_parameter: ...
```
Given the above, I would imagine the whole `protocol` block would be passed to your configuration modulation, creating some expanded list of configs, doing things like creating variations due to the "iterate".
I might then think that you could call `smqtk_core.configuration.from_config_dict(config_variation, Protocol.get_impls())` to get an instance of your protocol implementation type, and then run your `run_protocol` against it.

Reply:
Your strawman is more or less what happens now, except we're just passing in the result to the runtime invocation of `run_protocol()`, and there's no specific `protocol` section in the config.

Your approach would work, though I'm not yet entirely convinced of the benefits of using it yet.

You mentioned typechecking, which I don't think comes into play, except as a runtime check (TypeScript has runtime type predicates that can influence the typechecker's analysis using boolean paths through the code; not sure if Python typechecking has this feature). In the current situation, I believe you can add a type annotation to the `config` argument of a given `Protocol.run_protocol()` method, but this would be a preference of whoever is writing the protocol.

Your integrated SMQTK-centric approach may indeed be better than what we're doing, but I think it would be a good idea to discuss it live, in order for me to better understanding your thinking on the topic.